types/wlr_workspace_v1.c

```diff
@@ -0,0 +1,544 @@
+		struct wl_resource *resource) {
+	wl_resource_destroy(resource);
+}
+
+static void workspace_handle_activate(struct wl_client *client,
+		struct wl_resource *resource) {
+	struct wlr_workspace_handle_v1 *workspace = workspace_from_resource(resource);
+	if (!workspace) {
+		return;
+	}
+
+	workspace->pending |= WLR_WORKSPACE_HANDLE_V1_STATE_ACTIVE;
+}
+
+static void workspace_handle_deactivate(struct wl_client *client,
+		struct wl_resource *resource) {
+	struct wlr_workspace_handle_v1 *workspace = workspace_from_resource(resource);
+	if (!workspace) {
+		return;
+	}
+
+	workspace->pending &= ~WLR_WORKSPACE_HANDLE_V1_STATE_ACTIVE;
+}
+
+static void workspace_handle_resource_destroy(struct wl_resource *resource) {
+	wl_list_remove(wl_resource_get_link(resource));
+}
+
+static bool fill_array_from_workspace_state(struct wl_array *array,
+		uint32_t state) {
+	if (state & WLR_WORKSPACE_HANDLE_V1_STATE_ACTIVE) {
+		uint32_t *index = wl_array_add(array, sizeof(uint32_t));
+		if (index == NULL) {
+			return false;
+		}
+		*index = ZWLR_WORKSPACE_HANDLE_V1_STATE_ACTIVE;
+	}
+
+	return true;
+}
+
+static void workspace_handle_send_state_to_resource(
+		struct wlr_workspace_handle_v1 *workspace, struct wl_resource *resource) {
+	if (workspace->name) {
+		zwlr_workspace_handle_v1_send_name(resource, workspace->name);
+	}
+
+	if (workspace->coordinates.size > 0) {
+		zwlr_workspace_handle_v1_send_coordinates(resource,
+			&workspace->coordinates);
+	}
+
+	struct wl_array state;
+	wl_array_init(&state);
+	if (!fill_array_from_workspace_state(&state, workspace->server_state)) {
+		wl_resource_post_no_memory(resource);
+		wl_array_release(&state);
+		return;
+	}
+
+	zwlr_workspace_handle_v1_send_state(resource, &state);
+}
+
+static const struct zwlr_workspace_handle_v1_interface workspace_handle_impl = {
+	.destroy    = workspace_handle_destroy,
+	.activate   = workspace_handle_activate,
+	.deactivate = workspace_handle_deactivate,
+};
+
+void wlr_workspace_handle_v1_set_name(struct wlr_workspace_handle_v1 *workspace,
+		const char* name) {
+	free(workspace->name);
+	workspace->name = strdup(name);
+
+	struct wl_resource *tmp, *resource;
+	wl_resource_for_each_safe(resource, tmp, &workspace->resources) {
+		zwlr_workspace_handle_v1_send_name(resource, name);
+	}
+
+	workspace_manager_update_idle_source(workspace->group->manager);
+}
+
+void wlr_workspace_handle_v1_set_coordinates(
+		struct wlr_workspace_handle_v1 *workspace, struct wl_array *coordinates) {
+	wl_array_copy(&workspace->coordinates, coordinates);
+
+	struct wl_resource *tmp, *resource;
+	wl_resource_for_each_safe(resource, tmp, &workspace->resources) {
+		zwlr_workspace_handle_v1_send_coordinates(resource, coordinates);
+	}
+
+	workspace_manager_update_idle_source(workspace->group->manager);
+}
+
+void wlr_workspace_handle_v1_set_active(
+		struct wlr_workspace_handle_v1 *workspace, bool activate) {
+	if (activate) {
+		workspace->server_state |= WLR_WORKSPACE_HANDLE_V1_STATE_ACTIVE;
```

Comment:
I think we should change pending here, not server_state

Reply:
This is the function that the compositor calls, not the client. The client requests are `workspace_handle_activate/deactivate`

Reply:
But `server_state` not used outside here.

Reply:
The client may request a workspace to be focused. But the compositor may decide to ignore this request. Thus, there is one state which the client requested, and another state, that the compositor has set.

Reply:
Hmm. 
Commit method just change pending state to current? And clients change pending state directly?
I think we should prevent changing in this case via pending and current.
Because sync 3 state it's not obvious and easy.

Reply:
Client call request to change state
Compositor recive request and decide to change pending state or not.
Don't see any reasons for server state

Reply:
In the future we may have more states than just active/inactive so the `server_state` field will have more information that just active/inactive. In such cases, we will need to store them separately from pending/current (but yes, right now we could remove them. I decided to try to make the implementation a little bit future-proof because I expect a few more states like urgent).

Reply:
I expect too, but can't imagine the case when we want to separate server_state from current and pending state.
If compositor refuse to update state after request - it can just don't update pending.
If required some logic for complex change state by one request - it also can be performed via pending_state.

Reply:
I think you are misunderstanding what current and pending is. Pending is what the client has requested, but not yet committed. Once the client commits, that state becomes current - currently requested. So you can think of the `pending` field as a private implementation detail, compositors should not use it. They should use current only, to see what the client has requested.

And in server_state we save what the compositor has last requested too.

Reply:
-> And in server_state we save what the compositor has last requested too.
compositor requested from who?

Reply:
Why compositor not should use pending?
I see it like:
Client, Server:
- C: I want A
- S: -
- C: I want B
- S; -
- C: commit
- S: No, i don't give you A, but I'll give you B and C. *transorm pending to current and sends such events*

server_state seems like exceed duplicate of pending.

Reply:
I think we clarified why this is needed on IRC.

Reply:
Yes